### PR TITLE
[cairo] v2 support

### PR DESCRIPTION
**Pull request by planetmarshall**
Specify library name and version:  **cairo/**

The cairo package sets the name of the base pkg-config component file to "cairo" which renames the default parent component (the one containing all the sub-components).

The result is that dependent packages using the pkg-config generator and linking to "cairo" only get the base cairo component and not additional components such as `cairo-gobject`. This can result linking errors in dependent packages (specifically in the MSVC build of gtkmm which expects `cairo` to also include `cairo-gobject`)

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/cairo/all/conanfile.py [hunk 1]
```diff
@@ -222,11 +224,9 @@ def package(self):
         tools.remove_files_by_mask(self.package_folder, "*.la")
 
         self.copy("COPYING*", src=self._source_subfolder, dst="licenses", keep_path=False)
-        tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
-
+        files.rmdir(self, os.path.join(self.package_folder, "lib", "pkgconfig"))
 
     def package_info(self):
-        self.cpp_info.components["cairo_"].names["pkg_config"] = "cairo"
```
Comment (SpaceIm) [inline recipes/cairo/all/conanfile.py:229]: Please keep all these `pkg_config` names (you can translate them to `set_property("pkg_config_name", foo)`), so we can track these names are properly handled

To fix your issue, add a `pkg_config_name` at self.cpp_info level like `cairo-all-do-not-use`

File: recipes/cairo/meson/conanfile.py [hunk 2]
```diff
@@ -1,11 +1,15 @@
...
 import contextlib
 import glob
 import os
 
+from conan import ConanFile
+from conan.errors import ConanInvalidConfiguration
+from conan.tools import files, microsoft
+from conans import tools, Meson, VisualStudioBuildEnvironment
+
 required_conan_version = ">=1.38.0"
```
Comment (uilianries) [inline recipes/cairo/meson/conanfile.py:10]:
```suggestion
required_conan_version = ">=1.50.0"
```
new imports, new version.

File: recipes/cairo/all/conanfile.py [hunk 3]
```diff
@@ -1,8 +1,12 @@
...
-from conans.errors import ConanInvalidConfiguration
 import os
 import shutil
 
+from conan import ConanFile
+from conan.errors import ConanInvalidConfiguration
+from conan.tools import files, microsoft, scm
+from conans import AutoToolsBuildEnvironment, VisualStudioBuildEnvironment
+from conans import tools
+
```
Comment (uilianries) [inline recipes/cairo/all/conanfile.py:9]:
```suggestion

required_conan_version = ">=1.50.0"

```

**Comment by planetmarshall** [comment]
Failing due to missing binaries. See #11684

**Comment by planetmarshall** [comment]
This appears to have failed at the test package stage which is a bit confusing - surely it's just built the `cairo/1.17.4` package so why would it be missing?

File: recipes/cairo/meson/conanfile.py [hunk 4]
```diff
@@ -224,166 +226,119 @@ def package(self):
...
-            self.cpp_info.components["cairo-script-interpreter"].requires = ["cairo_"]
-
-        # binary tools
-        if self.options.with_zlib and self.options.with_png:
-            self.cpp_info.components["cairo_util_"].requires.append("expat::expat") # trace-to-xml.c, xml-to-trace.c
+        self.cpp_info.components["cairo_"].requires += list(base_requirements)
+        self.cpp_info.components["cairo_"].system_libs += list(base_system_libs)
 
     def package_id(self):
         self.info.requires["glib"].full_package_mode()
```
Comment (hesham-essam) [inline recipes/cairo/meson/conanfile.py:344]: While we're at it, maybe change this to:
```python
if not self.options.shared:
         self.info.requires["glib"].full_package_mode()
```
In the default cause (shared glib), I don't see a reason for using `full_package_mode`.

**Review by uilianries — APPROVED** [review]

File: recipes/cairo/all/conanfile.py [hunk 5]
```diff
@@ -252,36 +266,42 @@ def package_info(self):
...
             if self.options.with_xlib:
                 self.cpp_info.components["cairo_"].requires.extend(["xorg::x11", "xorg::xext"])
+
         if tools.is_apple_os(self.settings.os):
             self.cpp_info.components["cairo_"].frameworks.append("CoreGraphics")
 
-
         if self.settings.os == "Windows":
-            self.cpp_info.components["cairo-win32"].names["pkg_config"] = "cairo-win32"
+            self.cpp_info.components["cairo-win32"].set_property("pkg_config_name", "cairo-win32")
```
Comment (prince-chrismc) [inline recipes/cairo/all/conanfile.py:274]: Please leave the old names for v1 generators

File: recipes/cairo/meson/conanfile.py [hunk 6]
```diff
@@ -224,166 +232,120 @@ def package(self):
...
+        files.rmdir(self, os.path.join(self.package_folder, "lib", "pkgconfig"))
+        files.rm(self, "*.pdb", os.path.join(self.package_folder, "bin"))
 
     def package_info(self):
-        self.cpp_info.components["cairo_"].names["pkg_config"] = "cairo"
+        base_requirements = {"pixman::pixman"}
+        base_system_libs = {}
+
+        def add_component_and_base_requirements(component, requirements, system_libs=None):
+            self.cpp_info.components[component].set_property("pkg_config_name", component)
```
Comment (prince-chrismc) [inline recipes/cairo/meson/conanfile.py:243]: please leave the old `.names["pkg-config"]` for the v1 generators (we have not started removing them)

**Comment by prince-chrismc** [comment]
I think it's safe to rename the PR title, `cario: v2 support` since theres a lot of great work

File: recipes/cairo/all/conanfile.py (hunk 5 above)
Comment (planetmarshall) [reply recipes/cairo/all/conanfile.py:274]: Will the default name not be the name of the component anyway?

Comment (prince-chrismc) [reply recipes/cairo/all/conanfile.py:274]: We like to be explicit 😄 however I am sure this suggestion will change in the near future

**Review by prince-chrismc — APPROVED** [review]

File: recipes/cairo/meson/conanfile.py [hunk 7]
```diff
@@ -224,166 +230,137 @@ def package(self):
...
+            self.cpp_info.components[component].names["pkg_config"] = component
+            self.cpp_info.components[component].requires += ["cairo_"] + requirements
+            base_requirements.update(set(requirements))
+            if system_libs is not None:
+                self.cpp_info.components[component].system_libs += system_libs
+                base_system_libs.update(set(system_libs))
+
+        self.cpp_info.set_property("pkg_config_name", "cairo-all-do-no-use")
+        self.cpp_info.names["pkg_config"] = "cairo-all-do-no-use"
         self.cpp_info.components["cairo_"].libs = ["cairo"]
```
Comment (hesham-essam) [inline recipes/cairo/meson/conanfile.py:251]: Don't forget to add a pkg-config name to `cairo_`
```python
self.cpp_info.components["cairo_"].set_property("pkg_config_name", "cairo")
```
Many libraries search for the `cairo` pkgconfig file.

**Review by hesham-essam — CHANGES_REQUESTED** [review]

**Comment by hesham-essam** [comment]
@planetmarshall There has to be a component that just has the pkg-config name "cairo" since many libraries expect it to exist. These changes make it that no pkg-config file is named "cairo.pc".

File: recipes/cairo/all/conanfile.py [hunk 8]
```diff
@@ -146,18 +154,20 @@ def _configure_autotools(self):
         if self._autotools:
             return self._autotools
 
+        def boolean(value):
+            return "yes" if value else "no"
```
Comment (SpaceIm) [inline recipes/cairo/all/conanfile.py:158]: yes_no lambda was more elegant I believe, why this change?

File: recipes/cairo/meson/conanfile.py [hunk 9]
```diff
@@ -224,166 +230,140 @@ def package(self):
...
+        files.rm(self, "*.pdb", os.path.join(self.package_folder, "bin"))
 
     def package_info(self):
-        self.cpp_info.components["cairo_"].names["pkg_config"] = "cairo"
+        base_requirements = {"pixman::pixman"}
+        base_system_libs = {}
+
+        def add_component_and_base_requirements(component, requirements, system_libs=None):
+            self.cpp_info.components[component].set_property("pkg_config_name", component)
+            self.cpp_info.components[component].names["pkg_config"] = component
```
Comment (SpaceIm) [inline recipes/cairo/meson/conanfile.py:242]: Can be removed, `pkg_config` generator inherits from`pkg_config_name` property as well.

File: recipes/cairo/meson/conanfile.py [hunk 10]
```diff
@@ -224,166 +230,140 @@ def package(self):
...
+            base_requirements.update(set(requirements))
+            if system_libs is not None:
+                self.cpp_info.components[component].system_libs += system_libs
+                base_system_libs.update(set(system_libs))
+
+        self.cpp_info.set_property("pkg_config_name", "cairo-all-do-no-use")
+        self.cpp_info.names["pkg_config"] = "cairo-all-do-no-use"
+
+        self.cpp_info.components["cairo_"].set_property("pkg_config_name", "cairo")
+        self.cpp_info.components["cairo_"].names["pkg_config_name"] = "cairo"
```
Comment (SpaceIm) [inline recipes/cairo/meson/conanfile.py:253]:
```suggestion
```

useless if `pkg_config_name` is there

File: recipes/cairo/meson/conanfile.py (hunk 9 above)
Comment (SpaceIm) [inline recipes/cairo/meson/conanfile.py:242]:
```suggestion
```

Comment (prince-chrismc) [reply recipes/cairo/meson/conanfile.py:242]: I thought we broke the connection between v1 and v2 generators 🤔 was that only for CMake?

Since we have not removed it in other recipes, I dont see the harm in leaving it here as well

File: recipes/cairo/all/conanfile.py (hunk 8 above)
Comment (planetmarshall) [reply recipes/cairo/all/conanfile.py:158]: Rejected by the linter ( [PEP-8](https://peps.python.org/pep-0008/#programming-recommendations) - E731 in Flake8 ).

As for the naming, I think `boolean` is a better name for a function than `yes_no`.

File: recipes/cairo/meson/conanfile.py (hunk 10 above)
Comment (planetmarshall) [reply recipes/cairo/meson/conanfile.py:253]: Awaiting consensus from reviewers.

File: recipes/cairo/meson/conanfile.py (hunk 9 above)
Comment (planetmarshall) [reply recipes/cairo/meson/conanfile.py:242]: Awaiting consensus from reviewers.

File: recipes/cairo/meson/conanfile.py (hunk 10 above)
Comment (planetmarshall) [reply recipes/cairo/meson/conanfile.py:253]: ... although this should be `["pkg_config"].name` anyway. (typo)

File: recipes/cairo/all/conanfile.py [hunk 11]
```diff
@@ -90,23 +103,18 @@ def requirements(self):
     def build_requirements(self):
         if self._settings_build.os == "Windows" and not tools.get_env("CONAN_BASH_PATH"):
             self.build_requires("msys2/cci.latest")
-        if not self._is_msvc:
+        if not microsoft.is_msvc(self):
             self.build_requires("libtool/2.4.6")
             self.build_requires("pkgconf/1.7.4")
             self.build_requires("gtk-doc-stub/cci.20181216")
```
Comment (uilianries) [inline recipes/cairo/all/conanfile.py:109]:
```suggestion
            self.tool_requires("libtool/2.4.6")
            self.tool_requires("pkgconf/1.7.4")
            self.tool_requires("gtk-doc-stub/cci.20181216")
```
https://docs.conan.io/en/latest/migrating_to_2.0/recipes.html#requirements

**Review by prince-chrismc — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
